<a href="https://colab.research.google.com/github/subhra-15/Introduction-to-Machine-Learning/blob/master/Data%20Exploration%20and%20Preprocessing/Data_exploration_independent_numerical_variable.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Data exploration of independent numerical variable**

# **Treating outliers**
There are two ways of treating outliers:-

1.   deleting 
2.   imputing



#**Imputing**
in imputing process we add value based on mean, median or mode of the data in that column in place where there is no data

In [29]:
#importing the dataset
from google.colab import files
  
  
uploaded = files.upload()

Saving Raw_Housing_Prices2[1].csv to Raw_Housing_Prices2[1] (1).csv


In [30]:
import pandas as pd
import io
  
df = pd.read_csv(io.BytesIO(uploaded['Raw_Housing_Prices2[1].csv']))
df

ID  ... Lot Area after Renovation (in Sqft)
0      7129300520  ...                                5650
1      6414100192  ...                                7639
2      5631500400  ...                                8062
3      2487200875  ...                                5000
4      1954400510  ...                                7503
...           ...  ...                                 ...
21604   263000018  ...                                1509
21605  6600060120  ...                                7200
21606  1523300141  ...                                2007
21607   291310100  ...                                1287
21608  1523300157  ...                                1357

[21609 rows x 21 columns]

from the info we can see that there are intotal 21609 column but in some variable it is less than 21609 this represent the missing values

We have plenty of independent variable some of them are numeric in nature and some of them are continous in nature 
Based on this there are two kinds of imputation 

1.   Object variable imputation
2.   Continous variable imputation



# **Continous Variable imputation**
In this we will be imputing all the numeric variables

In [5]:
# combining all the numeric variable under a single list
number_columns=['No of Bathrooms','Flat Area (in Sqft)', 'Lot Area (in Sqft)','Area of the House from Basement (in Sqft)','Latitude','Longitude','Living Area after Renovation (in Sqft)']

for imputation we will bw using impute from sklearn

In [6]:
import sklearn 
from sklearn.impute import SimpleImputer
imputer= SimpleImputer(strategy='median')

# missing_value specifies how missing values are represented in our database by default it is np.nan
# strategy specifies how we impute the missing value
df[number_columns]=imputer.fit_transform(df[number_columns])

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21609 entries, 0 to 21608
Data columns (total 21 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   ID                                         21609 non-null  int64  
 1   Date House was Sold                        21609 non-null  object 
 2   Sale Price                                 21609 non-null  float64
 3   No of Bedrooms                             21609 non-null  int64  
 4   No of Bathrooms                            21609 non-null  float64
 5   Flat Area (in Sqft)                        21609 non-null  float64
 6   Lot Area (in Sqft)                         21609 non-null  float64
 7   No of Floors                               21609 non-null  float64
 8   Waterfront View                            21609 non-null  object 
 9   No of Times Visited                        21609 non-null  object 
 10  Condition of the House

There are two steps in fit_transform function:-

1. FITTING PHASE: in this phase the function calculate the median value w.r.t every column that we passed as a parameter and stores it

2. TRANSFORM PHASE: in this phase function does the actual action of locating the missing value and imputing it using median strategy

# **Object or Categorical Variable imputing**
In this we will be using mode strategy

In [8]:
imputer=SimpleImputer(strategy='most_frequent')
df['Zipcode'].shape
#as the column of zipcode have only row detail and column detail are missing so we need to add it

(21609,)

In [9]:
column=df['Zipcode'].values.reshape(-1,1)
column.shape
df['Zipcode']=imputer.fit_transform(column)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21609 entries, 0 to 21608
Data columns (total 21 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   ID                                         21609 non-null  int64  
 1   Date House was Sold                        21609 non-null  object 
 2   Sale Price                                 21609 non-null  float64
 3   No of Bedrooms                             21609 non-null  int64  
 4   No of Bathrooms                            21609 non-null  float64
 5   Flat Area (in Sqft)                        21609 non-null  float64
 6   Lot Area (in Sqft)                         21609 non-null  float64
 7   No of Floors                               21609 non-null  float64
 8   Waterfront View                            21609 non-null  object 
 9   No of Times Visited                        21609 non-null  object 
 10  Condition of the House

#**To convert numerical variable to categorical variables**

In [11]:
df['Zipcode']=df['Zipcode'].astype(object)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21609 entries, 0 to 21608
Data columns (total 21 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   ID                                         21609 non-null  int64  
 1   Date House was Sold                        21609 non-null  object 
 2   Sale Price                                 21609 non-null  float64
 3   No of Bedrooms                             21609 non-null  int64  
 4   No of Bathrooms                            21609 non-null  float64
 5   Flat Area (in Sqft)                        21609 non-null  float64
 6   Lot Area (in Sqft)                         21609 non-null  float64
 7   No of Floors                               21609 non-null  float64
 8   Waterfront View                            21609 non-null  object 
 9   No of Times Visited                        21609 non-null  object 
 10  Condition of the House

# **To convert a categorical variable to numerical variable**
The steps are as fallow:
1. Find out all the unique value in the categorical variable
2. replace the unique value with their corresponding numerical values

In [14]:
df['No of Times Visited'].unique()

array(['None', 'Thrice', 'Four', 'Twice', 'Once'], dtype=object)

In [15]:
# recoding them into numerical values using pandas.map function

mapping={
    'None':'0',
    'Once':'1',
    'Twice':'2',
    'Thrice':'3',
    'Four':'4'
}

#using pandas.map function pass mapping into it
df['No of Times Visited']=df['No of Times Visited'].map(mapping)

#checking the unique value
df['No of Times Visited'].unique()

array(['0', '3', '4', '2', '1'], dtype=object)

3. Another reason for variable transformation is to derive a new variable from existing variable or combining two variable to transform a variable so that it become more useful

   In the below codes we will be deriving two new variable from the variable 'Renovated Years'


*   Ever Renovated: this will be assigned 'yes' if the renovated year have non zero value otherwise it will be assigned 'no;
*   Year Since Renovation: if renovated year have non zero value then it will calculated the year since the house is last renovated



In [31]:
# in this we will be using where function from numpy
import numpy as np
df['Ever Renovated']= np.where(df['Renovated Year']==0,'No','Yes')

In [32]:
df['Purchase Year']=pd.DatetimeIndex(df['Date House was Sold']).year
df['Year Since Renovation']=np.where(df['Ever Renovated']=='Yes',abs(df['Purchase Year']-df['Renovated Year']),0)

In [33]:
df

ID Date House was Sold  ...  Purchase Year  Year Since Renovation
0      7129300520     14 October 2017  ...           2017                      0
1      6414100192    14 December 2017  ...           2017                     26
2      5631500400    15 February 2016  ...           2016                      0
3      2487200875    14 December 2017  ...           2017                      0
4      1954400510    15 February 2016  ...           2016                      0
...           ...                 ...  ...            ...                    ...
21604   263000018         14 May 2017  ...           2017                      0
21605  6600060120    15 February 2016  ...           2016                      0
21606  1523300141        14 June 2017  ...           2017                      0
21607   291310100     15 January 2016  ...           2016                      0
21608  1523300157     14 October 2017  ...           2017                      0

[21609 rows x 24 columns]

# **Deleting Some of the column**
For the tidyness of the data we will be deleting some of the column
such as
1. Purchase Year
2. Date House was Sold
3. Renovated Year

In [34]:
df.drop(columns=['Purchase Year','Date House was Sold','Renovated Year'],inplace=True)
df

ID  Sale Price  ...  Ever Renovated  Year Since Renovation
0      7129300520    221900.0  ...              No                      0
1      6414100192    538000.0  ...             Yes                     26
2      5631500400    180000.0  ...              No                      0
3      2487200875    604000.0  ...              No                      0
4      1954400510    510000.0  ...              No                      0
...           ...         ...  ...             ...                    ...
21604   263000018    360000.0  ...              No                      0
21605  6600060120    400000.0  ...              No                      0
21606  1523300141    402101.0  ...              No                      0
21607   291310100    400000.0  ...              No                      0
21608  1523300157    325000.0  ...              No                      0

[21609 rows x 21 columns]